rust/kernel/fs.rs

```diff
@@ -280,6 +303,15 @@ pub mod ro {
             unsafe { &*(*self.0.get()).i_sb.cast() }
         }
 
+        /// Returns the data associated with the inode.
+        pub fn data(&self) -> &T::INodeData {
+            let outerp = container_of!(self.0.get(), INodeWithData<T::INodeData>, inode);
```

Comment:
why is there no `unsafe` needed for `container_of!`?

Reply:
I think invocations of the `container_of!` macro should require `unsafe`. What error do you get when you use `T::INodeData`?

Reply:
It complains that `offset` must be a constant. When I try to make it a constant, it gives me the following error:

```
error[E0401]: can't use generic parameters from outer function
    --> rust/kernel/fs.rs:324:64
    |
307 | impl<T: FileSystem + ?Sized> INode<T> {
    |      - type parameter from outer function
...
324 |         let outerp = container_of!(self.0.get(), INodeWithData<T::INodeData>, inode);
    |                                                                ^^^^^^^^^^^^ use of generic parameter from outer function
```

I fixed it with `build_assert`.

See the discussion about `offset_of` here: https://github.com/Rust-for-Linux/linux/pull/1037#discussion_r1345266204